ui/src/variables/utils/hydrateVars.ts

```diff
@@ -212,16 +227,24 @@ const constVariableValues = (
 
   A node `b` is a descendant of `a` if there exists a path from `a` to `b`.
 
-  This will loop forever if it is run on a node within a cycle or leading to a
-  cycle.
+  Checks visited to prevent looping forever
 */
-const collectDescendants = (node: VariableNode, acc = []): VariableNode[] => {
+const collectDescendants = (
+  node: VariableNode,
+  visited: Set<VariableNode> = new Set()
+): VariableNode[] => {
+  let descendants = []
+
+  // reduce children
   for (const child of node.children) {
-    acc.push(child)
-    collectDescendants(child)
+    if (visited.has(child)) {
+      continue
+    }
+    visited.add(child)
+    descendants = descendants.concat(child, collectDescendants(child, visited))
```

Comment:
I feel like keeping track of two accumulators (`children`, `visited`) is pretty confusing... could this work like the `collectAncestors` method instead?

```
const collectDescendents = (
  node: VariableNode,
  acc = new Set()
): VariableNode[] => {
  for (const child of node.children) {
    if (!acc.has(child)) {
      acc.add(child)
      collectAncestors(child, acc)
    }
  }

  return [...acc]
}
```

Reply:
works for me.